### ISSUE TITLE
overcoming buffering

### ISSUE THREAD

**Issue by pete312** [ronf/asyncssh#226]
I'm trying to implement tailing ability in parallel in this fashion. 

```
command = 'strbuf -eL -oL ' + command
async with asyncssh.connect(host) as conn:
    await conn.run(command, stdout='/tmp/prun_%s.out' % host),
```

But the output seems to be buffered even if I provide the -u on the shebang line or even try to unbuffer the sys.stdout. Can you glide me on this?

I don't believe this is an issue with asyncssh so is there a better forum to ask questions like this ?

Thanks

**Comment by ronf** [comment]
Sure - happy to help.

As you said, AsyncSSH itself doesn't introduce any buffering of its own in this case. Any data written to the SSH channel is consumed immediately and written to the file you have specified in this case as stdout in the conn.run() call.

That leaves two possible sources of buffering. One of them is when running the command on the remote machine. I'm not familiar with the 'strbuf' command you are running there, but I'm guessing that it is intended to set stdout/stderr of the command being run to be line-buffered, so you're probably good there.

The other source of buffering is when AsyncSSH opens the file you are directing stdout to. The "-u" option when running Python won't help you here, as that only affects sys.stdout and sys.stderr opened when Python starts up. When you open files yourself, you have to specify the buffering to use.

In this case, it is AsyncSSH opening the file for you, but AsyncSSH provides a "bufsize" argument that you can use when calling conn.run() to set the buffering. However, one thing you'll need to know is that AsyncSSH always opens files in binary mode. It lets you process Unicode data, but it handles the conversion itself without relying on the file object since it supports may different types of I/O targets and not all of them support Unicode. So, you can't use line buffering, but you don't really need to. Since the data coming in will already be line-buffered, you can just disable buffering on the file writes, which you'd get by doing:

```python
command = 'strbuf -eL -oL ' + command
async with asyncssh.connect(host) as conn:
    await conn.run(command, stdout='/tmp/prun_%s.out' % host, bufsize=0),
```

This should make any data written by the remote process go out to the file immediately, allowing things like "tail" to work. I did a quick test here, and it worked for me. Let me know if you run into any problems with this.

**Comment by ronf** [comment]
Closing due to inactivity. If you have additional questions, feel free to open a new issue.

**Comment by pete312** [comment]
By the way this did work out. Thank you for the suggestion
